src/System.Private.CoreLib/shared/System/Index.cs

```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public readonly struct Index
+    {
+        private readonly int _value;
+
+        public int Value => _value < 0 ? ~_value : _value;
+        public bool FromEnd => _value < 0;
+
+        public Index(int value, bool fromEnd)
+        {
+            if (value < 0)
+            {
+                throw new ArgumentException(SR.ArgumentOutOfRange_NeedNonNegNum, nameof(value));
+            }
+
+            _value = fromEnd ? ~value : value;
```

Comment:
It may be more efficient to use the high bit as a flag rather than taking the one's complement of the whole number. It means both this method and the `Value` property getter can be branchless.

Reply:
I assume `fromEnd` is going to be constant `true` or `false` most of the time. This should be written such that the JIT is able to optimize this efficiently. For example, if I have code like this: 

```csharp
Span<int> a = new int[3];
Span<int> b = a[1..2];
```

The JITed code produced for this should have no bound checks in it.

Reply:
ok, per Jan's comment, I'll keep the current implementation and we should follow up with the JIT to ensure the optimization is happening.